src/coreclr/jit/lsraxarch.cpp

```diff
@@ -2328,47 +2328,113 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree)
 
                 const bool copiesUpperBits = HWIntrinsicInfo::CopiesUpperBits(intrinsicId);
 
+                unsigned resultOpNum = 0;
+                LIR::Use use;
+                if (LIR::AsRange(blockSequence[curBBSeqNum]).TryGetUse(intrinsicTree, &use))
+                {
+                    resultOpNum = intrinsicTree->GetResultOpNumForFMA(use.User(), op1, op2, op3);
+                }
+
                 // Intrinsics with CopyUpperBits semantics cannot have op1 be contained
                 assert(!copiesUpperBits || !op1->isContained());
 
-                if (op2->isContained())
-                {
-                    // 132 form: op1 = (op1 * op3) + [op2]
+                // Prioritize contained node by checking if any op is contained first.
+                // If none of them is contained, check if any op is regOptional.
+                bool     hasContainedOp = false;
+                unsigned containedOpNum = 0;
 
-                    tgtPrefUse = BuildUse(op1);
-
-                    srcCount += 1;
-                    srcCount += BuildOperandUses(op2);
-                    srcCount += BuildDelayFreeUses(op3, op1);
+                if (op1->isContained() || op2->isContained() || op3->isContained())
+                {
+                    hasContainedOp = true;
                 }
-                else if (op1->isContained())
+                if (hasContainedOp)
                 {
-                    // 231 form: op3 = (op2 * op3) + [op1]
-
-                    tgtPrefUse = BuildUse(op3);
-
-                    srcCount += BuildOperandUses(op1);
-                    srcCount += BuildDelayFreeUses(op2, op1);
-                    srcCount += 1;
+                    if (op1->isContained())
+                    {
+                        containedOpNum = 1;
+                    }
+                    else if (op2->isContained())
+                    {
+                        containedOpNum = 2;
+                    }
+                    else
+                    {
+                        assert(op3->isContained());
+                        containedOpNum = 3;
+                    }
                 }
                 else
                 {
-                    // 213 form: op1 = (op2 * op1) + [op3]
+                    if (op1->IsRegOptional())
+                    {
+                        containedOpNum = 1;
+                    }
+                    else if (op2->IsRegOptional())
+                    {
+                        containedOpNum = 2;
+                    }
+                    else
+                    {
+                        assert(op3->IsRegOptional());
+                        containedOpNum = 3;
+                    }
+                }
 
-                    tgtPrefUse = BuildUse(op1);
+                // Intrinsics with CopyUpperBits semantics must have op1 as target
+                if (containedOpNum == 1 && !copiesUpperBits)
```

Comment:
I think the rest of this logic could be simplified a bit in combination with the `containedOpNum = ...` checks above.

In particular, all the below are effectively doing the same thing, so you could conceptualize this instead as:
```
resultOpNum = ...;
containedOpNum = ...;

GenTree* emitOp1 = op1;
GenTree* emitOp2 = op2;
GenTree* emitOp3 = op3;

switch (resultOpNum)
{
    case 1:
    {
        // code
    }

    case 2:
    {
        // code
    }

    case 3:
    {    
        if (containedOpNum == 2)
        {
            std::swap(emitOp1, emitOp2);
        }

        std::swap(emitOp1, emitOp3);
        break;
    }

    default:
    {
        assert(resultOpNum == 0);
        // code
    }
}

tgtPrefUse = BuildUse(emitOp1);

srcCount += 1;
srcCount += BuildDelayFreeUses(emitOp1, emitOp2);
srcCount += emitOp3->isContained() ? BuildOperandUses(emitOp3) : BuildDelayFreeUses(emitOp1, emitOp3);
```

This all works because there are two things to think about.
1. The tree represents operands as input by the function, so `op1 = lhsMul`, `op2 = rhsMul`, `op3 = addend`
2. The `BuildUse` logic is dependent on how we might emit the instruction and so `emitOp1`, `emitOp2` and `emitOp3` (for lack of a better name) represent the actual `xmm1`, `xmm2`, and `xmm3` positions for `vfmadd###ps xmm1, xmm2, [xmm3]`. We always want to preference `xmm1/emitOp1` and we always want to contain `xmm3/emitOp3`.